mm2src/coins/tendermint/tendermint_tx_history_v2.rs

```diff
@@ -0,0 +1,754 @@
+        ctx.coin.set_history_sync_state(HistorySyncState::Finished);
+        Self::change_state(WaitForHistoryUpdateTrigger::new())
+    }
+}
+
+async fn ensure_storage_init<Storage>(wallet_ids: &[WalletId], storage: &Storage) -> MmResult<(), String>
+where
+    Storage: TxHistoryStorage,
+{
+    for wallet_id in wallet_ids.iter() {
+        if let Err(e) = storage.init(wallet_id).await {
+            return MmError::err(format!("{:?}", e));
+        }
+    }
+
+    Ok(())
+}
+
+#[async_trait]
+impl<Coin, Storage> State for TendermintInit<Coin, Storage>
+where
+    Coin: TendermintTxHistoryOps,
+    Storage: TxHistoryStorage,
+{
+    type Ctx = TendermintTxHistoryCtx<Coin, Storage>;
+    type Result = ();
+
+    async fn on_changed(self: Box<Self>, ctx: &mut Self::Ctx) -> StateResult<Self::Ctx, Self::Result> {
+        ctx.coin.set_history_sync_state(HistorySyncState::NotStarted);
+
+        if let Err(e) = ensure_storage_init(&ctx.coin.all_wallet_ids(), &ctx.storage).await {
+            return Self::change_state(Stopped::storage_error(e));
+        }
+
+        Self::change_state(FetchingTransactionsData::new(
+            ctx.coin.my_address().expect("my_address can't fail"),
+            ctx.coin.get_denom_and_ticker_map(),
+        ))
+    }
+}
+
+#[async_trait]
+impl<Coin, Storage> LastState for Stopped<Coin, Storage>
+where
+    Coin: TendermintTxHistoryOps,
+    Storage: TxHistoryStorage,
+{
+    type Ctx = TendermintTxHistoryCtx<Coin, Storage>;
+    type Result = ();
+
+    async fn on_changed(self: Box<Self>, ctx: &mut Self::Ctx) -> Self::Result {
+        log::info!(
+            "Stopping tx history fetching for {}. Reason: {:?}",
+            ctx.coin.ticker(),
+            self.stop_reason
+        );
+
+        let new_state_json = json!({
+            "message": format!("{:?}", self.stop_reason),
+        });
+
+        ctx.coin.set_history_sync_state(HistorySyncState::Error(new_state_json));
+    }
+}
+
+#[async_trait]
+impl TendermintTxHistoryOps for TendermintCoin {
+    async fn get_rpc_client(&self) -> MmResult<HttpClient, TendermintCoinRpcError> { self.rpc_client().await }
+
+    async fn query_htlc(&self, id: String) -> MmResult<QueryHtlcResponseProto, TendermintCoinRpcError> {
+        self.query_htlc(id).await
+    }
+
+    fn decimals(&self) -> u8 { self.decimals }
+
+    fn set_history_sync_state(&self, new_state: HistorySyncState) {
+        *self.history_sync_state.lock().unwrap() = new_state;
+    }
+
+    async fn all_balances(&self) -> MmResult<AllBalancesResult, TendermintCoinRpcError> { self.all_balances().await }
+
+    fn all_wallet_ids(&self) -> Vec<WalletId> {
+        let tokens = self.tokens_info.lock();
+        let mut ids = vec![];
+
+        ids.push(WalletId::new(self.ticker().replace('-', "_")));
+
+        for (ticker, _info) in tokens.iter() {
+            ids.push(WalletId::new(ticker.replace('-', "_")));
+        }
+
+        ids
+    }
```

Comment:
We can use a single wallet ID derived from platform ticker here. The RPC will return transfers of all assets - and we can save their data to a single table, even without knowing the actual ticker, decimals, etc. - this data can be recalculated on the fly when tx history is requested like here:
https://github.com/KomodoPlatform/atomicDEX-API/blob/6f73a21e176dfe975373e232c8d62fe8c53765b2/mm2src/coins/my_tx_history_v2.rs#L413-L416

So, we can store the information to DB using:
1. Denom as Token ID.
2. Store amounts in denom units. Recalculate on the fly using decimals when history is requested.
3. Denom can be also used as ticker - can be also replaced on the fly.

Reply:
> we can save their data to a single table

Do you think this is efficient? Current implementation has horizontal scale with partitioning tables for each ticker. Which will provide faster reads on tx_history.

Reply:
I suppose that a single address won't likely have such number of transactions that partitioning has a noticeable impact. How many transactions should address have for partitioning to be reasonable?

Reply:
> How many transactions should address have for partitioning to be reasonable?

This is depends on the db model of the table which I haven't check it detailed. If you have so many data with constraits in single table, even writing operations starts getting slower. I believe this is really rare case where a user will have thousands of transactions on single address. But if we want to make long-term scalable solutions, this is the way we should implement. What's the advantage inserting all transactions to single table? If there is no advantage I think we should keep it this way.

Reply:
> Store amounts in denom units. Recalculate on the fly using decimals when history is requested.

Shouldn't we do such operations in one-time processes? I think mapping values on the fly even increases the complexity of the workflow. I can simply calculate this on tx-history side(it will also be calculated once, not on every history request).